struct.rkt

```diff
@@ -13,6 +16,31 @@
 
 (struct Decl (filename id left right) #:transparent)
 
+;; optional arguments
+
+(define undef-object (gensym 'undef))
+
+(define (undef? x)
+  (eq? x undef-object))
+
+(define (undef/c pred?)
+  (λ (x)
+    (or/c (undef? x) (pred? x))))
+
+(provide undef?
+         undef/c
+         undef-object)
+
+;; request error
+
+(define request-err-object (gensym 'request-error))
+
+(define (request-err? x)
+  (eq? x request-err-object))
+
+(provide request-err-object
+         request-err?)
```

Comment:
What we need `gensym` to protect it from outside?

If we really need this, I also recommend another implementation which is based on reference:
```racket
(struct request-err ())

(define request-err-object (request-err))
(define (request-err? e)
  (eq? request-err-object e))
```

Reply:
I think there is no much difference here.

My initial idea was to use structs.
Then I remembered `gensym` which allowed me to avoid define additional structs. It generates an uninterned symbol so it's safe to compare with it.

Reply:
Oh, I make a measure, symbol comparison got less cpu and gc time. Just mark this resolved.

Appendix test code

```rkt
#lang racket

(define N 100000000)

(struct request-err-ty ())

(define request-err-object (request-err-ty))
(define (request-err? e)
  (eq? request-err-object e))

(time
 (for ([i N])
   (request-err? request-err-object)
   (request-err? 'abc)))

(define request-err-object2 (gensym 'request-error))

(define (request-err2? x)
  (eq? x request-err-object2))

(time
 (for ([i N])
   (request-err2? request-err-object2)
   (request-err2? 'abc)))
```